src/hotspot/share/runtime/vframeArray.cpp

```diff
@@ -308,7 +314,9 @@ void vframeArrayElement::unpack_on_stack(int caller_actual_parameters,
     top = iframe()->previous_monitor_in_interpreter_frame(top);
     BasicObjectLock* src = _monitors->at(index);
     top->set_obj(src->obj());
+    assert(ObjectSynchronizer::current_thread_holds_lock(thread, Handle(thread, src->obj())), "should be held, before move_to");
```

Comment:
Same comment as above, may need to check for null obj.

Reply:
Not sure how `obj` can be null in this code. ???

Reply:
That is fetching from an index in the MonitorChunk* _monitors, so if we recognise null in element->fill_in() when populating MonitorChunk*, seems good to recognise it here in vframeArrayElement::unpack_on_stack()?

Reply:
I can follow that logic but ... if it is null then what is this code actually doing? We have determined that the frame does contain locked monitors and so we are transferring them across. How can such a locked monitor have a null object?

Reply:
I assume it's only for the `fill_in` `realloc_failures` case.  But you're right, it doesn't seem very useful.  It's just going to look like an unlocked monitor slot in the interpreter frame.  We could consider skipping these in `fill_in`, then they won't show up later in `unpack_on_stack`().

Reply:
fill_in() has previously OK with seeing monitor->owner() == nullptr
so it's already setting dest->set_obj(null) under some conditions.

I see we can handle the null separately and simplify the asserts there.

vframeArrayElement::unpack_on_stack()
Still might retrieve a null, so the asserts there keep the guard against doing the owner check -- I'm not sure if they won't show up there - it loops over the number of elements in the MonitorChunk* so it should see them all?

Reply:
(marking as resolved, null checks are back in these asserts, there could be a null there from the is_scalar_replaced case, and possibly if realloc_failures is true)